File: frontend/src/api/rabbitGrowth.ts
/**
 * Rabbit Growth & Feed API (Module 17) — weight history, growth analysis, feeding
 * (which reuses the platform Inventory module) and feed efficiency. All growth and
 * FCR math is the backend's; the client renders honesty-labelled figures only.
 */
import apiClient from "./client";
import type { Figure } from "./rabbit";

type APISuccess<T> = { data: T; success: true };
type APIList<T> = { data: T[]; success: true; meta: { total: number; page: number; limit: number; pages: number } };
const base = (farmId: string) => `/farms/${farmId}/rabbit`;

export interface Weight {
  id: string; rabbit_id: string; recorded_on: string; weight_g: string; age_days: number | null; notes: string | null;
}
export interface GrowthAnalysis {
  rabbit_id: string;
  internal_ref: string;
  analysis: Record<string, Figure>;
  series: Array<{ recorded_on: string; weight_g: number | null; age_days: number | null; gain_from_prev_g: number | null; adg_from_prev: number | null }>;
  herd_percentile_pct: Figure;
}
export interface FeedRecord {
  id: string; rabbit_id: string | null; cage_id: string | null; inventory_item_id: string | null;
  inventory_movement_id: string | null; feed_type: string | null; quantity_kg: string;
  fed_on: string; cost: string | null; currency: string | null; supplier: string | null; notes: string | null;
}

export async function listWeights(farmId: string, rabbitId: string) {
  const { data } = await apiClient.get<APISuccess<Weight[]>>(`${base(farmId)}/rabbits/${rabbitId}/weights`);
  return data.data;
}
export async function recordWeight(farmId: string, rabbitId: string, body: { recorded_on: string; weight_g: number; notes?: string }) {
  const { data } = await apiClient.post<APISuccess<Weight>>(`${base(farmId)}/rabbits/${rabbitId}/weights`, body);
  return data.data;
}
export async function getGrowthAnalysis(farmId: string, rabbitId: string) {
  const { data } = await apiClient.get<APISuccess<GrowthAnalysis>>(`${base(farmId)}/rabbits/${rabbitId}/growth`);
  return data.data;
}
export async function listFeed(farmId: string, params: Record<string, unknown> = {}) {
  const { data } = await apiClient.get<APIList<FeedRecord>>(`${base(farmId)}/feed`, { params });
  return data;
}
export async function recordFeed(farmId: string, body: Record<string, unknown>) {
  const { data } = await apiClient.post<APISuccess<FeedRecord>>(`${base(farmId)}/feed`, body);
  return data.data;
}
export async function getFeedSummary(farmId: string, rabbitId?: string) {
  const { data } = await apiClient.get<APISuccess<{ rabbit_id: string | null; summary: Record<string, Figure> }>>(
    `${base(farmId)}/feed/summary`, { params: rabbitId ? { rabbit_id: rabbitId } : {} });
  return data.data;
}

File: backend/app/api/v1/endpoints/rabbit_growth.py
"""
Greena — Rabbit Growth & Feed API (Module 17, Milestone 4).

Weight recording + growth analysis, and feeding (which reuses the platform
Inventory module). Every route is farm-scoped and permission-guarded.

Route map (prefix /farms/{farm_id}/rabbit):
  Growth / weight
    POST   /rabbits/{id}/weights          record a weight (RABBIT_WEIGHT_LOG)
    GET    /rabbits/{id}/weights          list weights (RABBIT_VIEW)
    GET    /rabbits/{id}/growth           growth analysis (RABBIT_VIEW)
  Feed (Inventory reuse)
    POST   /feed                          record a feeding (RABBIT_FEED_RECORD)
    GET    /feed                          list feeding (RABBIT_FEED_VIEW)
    GET    /feed/summary                  feed summary + FCR (RABBIT_FEED_VIEW)
"""

from math import ceil
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status

from app.core.permissions import Permission, require_permission
from app.dependencies import CurrentUser, DBSession, require_farm_access
from app.schemas.base import ListResponse, PaginationMeta, SuccessResponse
from app.schemas.rabbit import (
    FeedRecordCreate,
    FeedRecordResponse,
    WeightCreate,
    WeightResponse,
)
from app.services import rabbit_feed_service as feed_svc
from app.services import rabbit_growth_service as growth_svc

router = APIRouter(prefix="/farms/{farm_id}/rabbit", tags=["Rabbit Growth & Feed"])


# ── Weight & growth ──────────────────────────────────────────────────────────────

@router.post("/rabbits/{rabbit_id}/weights", response_model=SuccessResponse[WeightResponse], status_code=status.HTTP_201_CREATED)
async def record_weight(
    farm_id: str, rabbit_id: UUID, body: WeightCreate, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_WEIGHT_LOG)),
):
    farm, _ = access
    row = await growth_svc.record_weight(db, farm, rabbit_id, body, current_user)
    return SuccessResponse(data=WeightResponse.model_validate(row))


@router.get("/rabbits/{rabbit_id}/weights", response_model=SuccessResponse[list[WeightResponse]])
async def list_weights(
    farm_id: str, rabbit_id: UUID, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_VIEW)),
):
    farm, _ = access
    rows = await growth_svc.list_weights(db, farm.id, rabbit_id)
    return SuccessResponse(data=[WeightResponse.model_validate(r) for r in rows])


@router.get("/rabbits/{rabbit_id}/growth", response_model=SuccessResponse[dict])
async def growth_analysis(
    farm_id: str, rabbit_id: UUID, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_VIEW)),
):
    farm, _ = access
    return SuccessResponse(data=await growth_svc.growth_analysis(db, farm.id, rabbit_id))


# ── Feed (reuses platform Inventory) ─────────────────────────────────────────────

@router.post("/feed", response_model=SuccessResponse[FeedRecordResponse], status_code=status.HTTP_201_CREATED)
async def record_feeding(
    farm_id: str, body: FeedRecordCreate, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_FEED_RECORD)),
):
    farm, _ = access
    row = await feed_svc.record_feeding(db, farm, body, current_user)
    return SuccessResponse(data=FeedRecordResponse.model_validate(row))


@router.get("/feed", response_model=ListResponse[FeedRecordResponse])
async def list_feeding(
    farm_id: str, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_FEED_VIEW)),
    rabbit_id: UUID | None = Query(None),
    cage_id: UUID | None = Query(None),
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
):
    farm, _ = access
    rows, total = await feed_svc.list_feeding(
        db, farm.id, rabbit_id=rabbit_id, cage_id=cage_id, limit=limit, offset=(page - 1) * limit
    )
    return ListResponse(
        data=[FeedRecordResponse.model_validate(r) for r in rows],
        meta=PaginationMeta(total=total, page=page, limit=limit, pages=ceil(total / limit) if total else 1),
    )


@router.get("/feed/summary", response_model=SuccessResponse[dict])
async def feed_summary(
    farm_id: str, db: DBSession, current_user: CurrentUser,
    access: tuple = Depends(require_farm_access()),
    _perm=Depends(require_permission(Permission.RABBIT_FEED_VIEW)),
    rabbit_id: UUID | None = Query(None),
):
    farm, _ = access
    return SuccessResponse(data=await feed_svc.feed_summary(db, farm.id, rabbit_id))
